src/ripple/app/main/NodeStoreScheduler.cpp

```diff
@@ -48,16 +46,21 @@ void NodeStoreScheduler::onChildrenStopped ()
 void NodeStoreScheduler::scheduleTask (NodeStore::Task& task)
 {
     ++m_taskCount;
-    m_jobQueue->addJob (
+    if (!m_jobQueue->addJob (
         jtWRITE,
         "NodeObject::store",
-        [this, &task] (Job&) { doTask(task); });
+        [this, &task] (Job&) { doTask(task); }))
+    {
+        // Job not added, presumably because we're shutting down.
+        // Recover by executing the task synchronously.
+        doTask (task);
+    }
 }
 
 void NodeStoreScheduler::doTask (NodeStore::Task& task)
 {
     task.performScheduledTask ();
-    if ((--m_taskCount == 0) && isStopping())
+    if ((--m_taskCount == 0) && isStopping() && areChildrenStopped())
```

Comment:
I agree with your assessment on both counts:

1. The diagram is useful and should be somewhere in the source code even if it gets stale.  I've toyed with a couple of places to leave the diagram.  Application.cpp is where a lot of this structure gets built.  So somewhere in there almost makes sense.  But I'm leaning toward putting the diagram in Stoppable.h.  This is a little weird.  All of the elements of the diagram are `Stoppable`, but `Stoppable` itself is ignorant of the teaming classes below that it serves.  Nevertheless, I think putting the diagram in Stoppable.h improves the discoverability of the diagram.  What are your thoughts?

2. I agree with your assessment that, if there are no bugs, it should never be possible for `doTask` to call `stopped()`.  I'm hesitant to make such a change in this particular pull request.  However I'd like to add a comment in the `doTask()` source including your assessment.  This might encourage someone in the future to study the situation and potentially remove the call to `stopped()`.